### ISSUE TITLE
Possible issue with incoming changes for banking system (SMS auth)

### ISSUE THREAD

**Issue by KptChaos** [jwilk/mbank-cli#70]
Possible issue with incoming changes for banking system (SMS auth) Since 14 September additional SMS auth during login into account will be required which means mbank-cli may become useless. Is there any solutution?

**Comment by jwilk** [comment]
The minimum plan is to make mbank-cli ask for the SMS password on tty. This is of course only a sufficient solution for people who use the program interactively.

Making mbank-cli work non-interactively is a much tougher problem. Here are some ideas:

* Require the user to have the phone connected to the computer and automatically read the SMS messages (via [Gammu](https://wammu.eu/gammu/) or something).

* Reverse-engineer the mobile app and re-implement it as part of mbank-cli. (No idea if this is feasible, and anyway I don't plan to work on this.)

* Ask mBank to use standardized 2FA mechanisms, such as [TOTP](https://en.wikipedia.org/wiki/Time-based_One-time_Password_Algorithm) or [U2F](https://en.wikipedia.org/wiki/U2F), and then implement them in mbank-cli. (Unlikely to be feasible.)

* The SMS password will be required only(?) for login, so perhaps we could make mbank-cli keep the user logged in forever somehow?

**Comment by KptChaos** [comment]
The first idea... is the best. The biggest problem is... to read SMS. Usually bank allows to specify 1 phone number to each user.
If you are LTD or something, you can have multiply users and multiply phone numbers to authorize transaction (if you SMS auth) I don't know it its possible to have few logins (entities) if you;'re self employed for example... Despite that. It may be some solution using gnokii or gammu (I use the gnokii) that way or the other I believe for most people it will be pain in the ass. Besides it may be additional "point of failure" (phone conected to PC may just hang)  When it comes to other ideas... Sorry, but "I don't see it" Impossible to do it. .

**Comment by jwilk** [comment]
> The SMS password will be required only(?) for login, so perhaps we could make mbank-cli keep the user logged in forever somehow?

Fun fact: PSD2 requires that the session expire after 5 minutes (or less).
Source: <https://eur-lex.europa.eu/eli/reg_del/2018/389/oj>, §4.3d.

**Comment by chilek** [comment]
some interesting info:
https://www.mbank.pl/mbank-news/co-nowego/od-14-wrzesnia-duzo-zmian-w-banku-nie-daj-sie-zaskoczyc.html
and pt 3.

**Comment by netmaniac** [comment]
With PSD2 implementation banks were required to create API. https://developer.api.mbank.pl/ I don't know what are requirements to get access. Looks like it will require something like this  https://www.elektronicznypodpis.pl/zastosowania/certyfikaty-do-psd2/ 

Price ~1200 PLN (~300 EUR)

**Comment by netmaniac** [comment]
Czy każdy może skorzystać z mBank API:
Niestety nie każdy ma taką możliwość. Dostęp do mBank API może mieć tylko przedsiębiorca który świadczy usługi dostępu do rachunku lub inicjacji płatności. Dodatkowo musi dostać licencję na świadczenie takich usług od krajowego nadzorcy (w Polsce jest to KNF). Zgodnie z regulacją europejską, aby mieć dostęp do mBank API PSD2 trzeba mieć status TPP (Third Party Provider).

**Comment by jwilk** [comment]
> https://www.mbank.pl/mbank-news/co-nowego/od-14-wrzesnia-duzo-zmian-w-banku-nie-daj-sie-zaskoczyc.html
> and pt 3.

They say you could make add make your device "trusted". Sounds like a perfect fit for mbank-cli. But it's not a complete solution, because they will occasionally ask for the second factor even on trusted devices. :-/

Another interesting piece information is that there will be another application ("mBank Token") just for mobile authentication. This might be easier to reverse-engineer than the full-blown mobile app.

**Comment by jwilk** [comment]
ETA for interactive SMS password prompt: this evening.

**Comment by chilek** [comment]
How can I send you a little tip?

**Comment by jwilk** [comment]
> How can I send you a little tip?

Please drop me an email. :-)

**Comment by jwilk** [comment]
I've added initial support for interactive SMS auth in git master (115bae2cd23cd80d883041c39c2159045d6fd7ee). Early testing is welcome!
This is not very robust at the moment. In particular, it doesn't handle misspelt passwords gracefully, so please don't make mistakes. :-)

**Comment by chilek** [comment]
Updated to current master and I have:
```
$ ./mbank-cli activate-profile business
Scraping error: login.2fa.init.auth-mode: NAM at /opt/chilan/mbank-cli line 1215.
	main::do_login() called at ./mbank-cli line 2761
Please file a bug at <https://github.com/jwilk/mbank-cli/issues>.
$
```
At the same time I got mbank mobile authorization request in my phone. I don't use sms one-time password. I suppose that OTP form is different than you assumed for SMS OTP :(

**Comment by jwilk** [comment]
>```
>Scraping error: login.2fa.init.auth-mode: NAM at /opt/chilan/mbank-cli line 1215.
>```

Yeah, 2FA via the mobile app is not implemented yet.
I've opened #72 for this, and made the error message a bit more helpful (04a1a81eab214c59a1ced0374131a9caf52bc799).

**Comment by jwilk** [comment]
> The SMS password will be required only(?) for login, so perhaps we could make mbank-cli keep the user logged in forever somehow?

This seems to work, at least for small values of "forever".
I ran mbank-cli in a loop with a 3–4 minutes delay between iterations, and was able to keep the session alive for a few hours.

(I don't plan to add any code to mbank-cli to extend the session lifetime at the moment. It looks like users who want that can easily implement it themselves.)

**Comment by jwilk** [comment]
> Require the user to have the phone connected to the computer and automatically read the SMS messages (via [Gammu](https://wammu.eu/gammu/) or something)

There's now an (undocumented and experimental) option to read SMS messages from disk:
```
SMSInbox /path/to/sms/inbox
```
You could combine that with [gammu-smsd](https://wammu.eu/docs/manual/smsd/) (with files backend) to have non-interactive login.
I suspect this is going to be quite brittle in practice, so I wouldn't recommend it for unattended use.

Most likely only Polish SMS messages are parsed correctly at the moment. Patches for Slovak and Czech are welcome.

**Comment by jwilk** [comment]
> I've added initial support for interactive SMS auth

I've just released [mbank-cli 2.0](https://github.com/jwilk/mbank-cli/releases/tag/2.0), which includes aforementioned changes.

If you run into any bugs, or badly need other authentication features, please file separate issues.
